### PR TITLE
feat: `TransOrd` and `BEqOrd` + instances

### PR DESCRIPTION
**Pull request by digama0**
This addresses a current hole in the API with respect to the RBMap lemmas. The issue is that most RBMap lemmas are written in terms of some `cmp` function satisfying `TransCmp`, but there is no way to get from `cmp x y = .eq` to `x = y` when the key type is a builtin. We solve this using two classes (and two `abbrev`s):

```lean
/-- `TransCmp cmp` asserts that `cmp` induces a transitive relation. -/
class TransCmp (cmp : α → α → Ordering) extends OrientedCmp cmp : Prop where
  /-- The comparator operation is transitive. -/
  le_trans : cmp x y ≠ .gt → cmp y z ≠ .gt → cmp x z ≠ .gt

/-- `BEqCmp cmp` asserts that `cmp x y = .eq` and `x == y` coincide. -/
class BEqCmp [BEq α] (cmp : α → α → Ordering) : Prop where
  /-- `cmp x y = .eq` holds iff `x == y` is true. -/
  cmp_iff_beq : cmp x y = .eq ↔ x == y

/-- `TransOrd α` asserts that the `Ord` instance satisfies `TransCmp`. -/
abbrev TransOrd (α) [Ord α] := TransCmp (α := α) compare

/-- `BEqOrd α` asserts that the `Ord` and `BEq` instances are coherent via `BEqCmp`. -/
abbrev BEqOrd (α) [BEq α] [Ord α] := BEqCmp (α := α) compare
```
The `abbrev`s allow stating a property directly about the `Ord` instance on a type. The combination of `TransOrd` and `BEqOrd`, along with `LawfulBEq`, suffice to establish that the type is linearly ordered by the `Ord` instance, so `compare x y = .eq <-> x = y` holds.

Most of the actual work in this PR is proving that every type from `Init` and `Lean` (except for a few oddballs like `JsonNumber` that we can revisit later) with an `Ord` instance also has a `TransOrd` and `BEqOrd` instance.

This also needs a corresponding mathlib PR, not because of breakage but to have instances of `TransOrd` and `BEqOrd` for `LinearOrder` as well.

### CODE REVIEW

**Review by fgdorais — APPROVED** [review]

File: Std/Classes/Order.lean [hunk 1]
```diff
@@ -88,6 +107,247 @@ theorem cmp_congr_right [TransCmp cmp] (yz : cmp y z = .eq) : cmp x y = cmp x z
...
+/-- `LECmp cmp` asserts that `cmp x y ≠ .gt` and `x ≤ y` coincide. -/
+class LECmp [LE α] (cmp : α → α → Ordering) extends OrientedCmp cmp : Prop where
+  /-- `cmp x y = .eq` holds iff `x < y` is true. -/
+  cmp_iff_le : cmp x y ≠ .gt ↔ x ≤ y
+
+theorem LTCmp.cmp_iff_ge [LE α] [LECmp (α := α) cmp] : cmp x y ≠ .lt ↔ y ≤ x := by
+  rw [← OrientedCmp.cmp_ne_gt, LECmp.cmp_iff_le]
+
+/-- `LawfulCmp α` asserts that the `Ord`, `LE`, `LT` instances are all coherent with each other,
+describing a strict weak order (a linear order except for antisymmetry). -/
```
[the remark this answers is not in the document]
Comment (digama0) [reply Std/Classes/Order.lean:135]:
```suggestion
/-- `LawfulCmp cmp` asserts that the `LE`, `LT`, `BEq` instances are all coherent with each other 
and with `cmp`, describing a strict weak order (a linear order except for antisymmetry). -/
```

**Comment by fgdorais** [comment]
Tangential: Core's `compareOfLessAndEq` is not the best choice since it requires two decidable relations, including decidable equality. 

The following two alternatives are objectively better:
```lean
def compareOfLE {α} [LE α] [DecidableRel (α:=α) (.≤.)] (x y : α) : Ordering :=
  if x ≤ y then if y ≤ x then .eq else .lt else .gt
```
which is designed to work with `LECmp`.
```lean
def compareOfLT {α} [LT α] [DecidableRel (α:=α) (.<.)] (x y : α) : Ordering :=
  if x < y then .lt else if y < x then .gt else .eq
```
which is designed to work with `LTCmp`.

None of these should be considered "canonical" since they *all* lead to diamond issues.

**Comment by digama0** [comment]
`compareOfLT` is defined in mathlib by the name `cmp` too.

As far as "objectively better", this also depends on the performance of the two operations. I can imagine situations where < is more expensive than ==, in which case `compareOfLessAndEq` might perform better than `compareOfLT`.

Re: canonical, this is true, we certainly can't always infer `cmp` from `LT` or `BEq` or both. Besides the diamond issues, there is also the case that we can implement `cmp` directly and more efficiently (which is not uncommon, especially in derived `Ord` instances). Primitive types usually don't come with `cmp` built in though, because the hardware (or at least C) don't provide native three-way comparison mechanisms. (But see [C++](https://brevzin.github.io/c++/2019/07/28/comparisons-cpp20/)...) So in these cases we use `compareOfLessAndEq` or its friends to produce a `cmp` instance, and the theorems here allow us to prove that these instances are suitably well behaved.

**Comment by fgdorais** [comment]
By "objectively better", I only meant that neither requires decidable equality. Sorry that wasn't clear.

**Comment by fgdorais** [comment]
Re: "canonical", I completely agree. I would only add that vectorized comparisons are commonly available in hardware, they often use `cmp` rather than plain ordering relations.

**Comment by semorrison** [comment]
There's a conflict, unfortunately, but otherwise I think we can merge.

**Pull request merged**